ui/src/app/pages/workspace/workspace-wrapper.tsx

```diff
@@ -134,9 +134,92 @@ const MultiRegionWorkspaceNotification = () => {
   );
 };
 
+interface UpdateRuntimeStoresProps {
+  workspaceNamespace: string;
+  pollAborter: AbortController;
+  setPollAborter: (ac: AbortController) => void;
+}
+const updateRuntimeStores = async ({
+  workspaceNamespace,
+  pollAborter,
+  setPollAborter,
+}: UpdateRuntimeStoresProps) => {
+  runtimeDiskStore.set({
+    workspaceNamespace,
+    gcePersistentDisk: undefined,
+  });
+  runtimeStore.set({
+    workspaceNamespace,
+    runtime: undefined,
+    runtimeLoaded: false,
+  });
+  pollAborter.abort();
+  const newPollAborter = new AbortController();
+  setPollAborter(newPollAborter);
+
+  try {
+    await LeoRuntimeInitializer.initialize({
+      workspaceNamespace,
+      pollAbortSignal: newPollAborter.signal,
+      maxCreateCount: 0,
+      maxResumeCount: 0,
+    });
+  } catch {
+    // Ignore InitialRuntimeNotFoundError.
+    // Ignore ExceededActionCountError. This is thrown when the runtime doesn't exist, or
+    // isn't started. Both of these scenarios are expected, since we don't want to do any lazy
+    // initialization here.
+    // Also ignore LeoRuntimeInitializationAbortedError - this is expected when navigating
+    // away from a page during a poll.
+    // Ideally, we would handle or log errors except the ones listed above.
+  }
+};
+
+interface GetWorkspaceProps {
+  wsid: string;
+  setWorkspace: (w: Workspace) => void;
+  navigate: (path: string[]) => void;
+}
+const getWorkspaceAndUpdateStores = async ({
+  workspaceNamespace,
+  wsid,
+  setWorkspace,
+  navigate,
+  pollAborter,
+  setPollAborter,
+}: GetWorkspaceProps & UpdateRuntimeStoresProps) => {
+  try {
+    const { workspace, accessLevel } = await workspacesApi().getWorkspace(
+      workspaceNamespace,
+      wsid
+    );
+    currentWorkspaceStore.next({
+      ...workspace,
+      accessLevel,
+    });
+    updateRuntimeStores({
+      workspaceNamespace: workspace.namespace,
+      pollAborter,
+      setPollAborter,
+    });
+    setWorkspace(workspace);
+  } catch (ex) {
+    if (ex.status === 403) {
+      navigate(['/workspaces']);
+    }
+  }
+};
+
 export const WorkspaceWrapper = ({ hideSpinner }) => {
   const params = useParams<MatchParams>();
   const { ns, wsid } = params;
+  const routeData = useStore(routeDataStore);
+  const [navigate] = useNavigation();
+
+  const [pollAborter, setPollAborter] = useState(new AbortController());
+  const [workspace, setWorkspace] = useState<Workspace>(undefined);
+  const [showNewCtNotification, setShowNewCtNotification] = useState(false);
+
   useEffect(() => {
```

Comment:
this component has a weirdly high number of `useEffect`s which I will not tackle in this PR, but look into soon